Wrappers/Python/ccpi/optimisation/functions/Function.py

```diff
@@ -23,55 +23,493 @@
+        
+        super(SumFunctionScalar, self).__init__(function, ConstantFunction(constant))        
+        self.constant = constant
+        self.function = function
+        
+    def convex_conjugate(self,x):
+        
+        r""" Returns the convex conjugate of a :math:`(F+scalar)`
+                
+        .. math:: (F+scalar)^{*}(x^{*}) = F^{*}(x^{*}) - scalar
+                        
+        """        
+        return self.function.convex_conjugate(x) - self.constant
+    
+    def proximal(self, x, tau, out=None):
+        
+        """ Returns the proximal operator of :math:`F+scalar`
+
+        .. math:: \mathrm{prox}_{\tau (F+scalar)}(x) = \mathrm{prox}_{\tau F}
+                        
+        """                
+        return self.function.proximal(x, tau, out=out)        
+            
+    def function(self):       
+       return self.function    
+
+class ConstantFunction(Function):
+    
+            
+    r""" ConstantFunction: :math:`F(x) = constant, constant\in\mathbb{R}`         
+        
+    """
+    
+    def __init__(self, constant = 0):
+        
+        super(ConstantFunction, self).__init__(L=0)
+        
+        if not isinstance (constant, Number):
+            raise TypeError('expected scalar: got {}'.format(type(constant)))
+                
+        self.constant = constant
+              
+    def __call__(self,x):
+        
+        """ Returns the value of the function, :math:`F(x) = constant`"""
+        return self.constant
+        
+    def gradient(self, x, out=None):
+        
+        """ Returns the value of the gradient of the function, :math:`F'(x)=0`"""       
+        
+        return ZeroOperator(x.geometry).direct(x, out=out)
+    
+    def convex_conjugate(self, x):
+                        
+        r""" The convex conjugate of constant function :math:`F(x) = c\in\mathbb{R}` is
+        
+        .. math:: 
+            F(x^{*}) 
+            =
+            \begin{cases}
+                -c, & if x^{*} = 0\\
+                \infty, & \mbox{otherwise}
+            \end{cases}
+                                                          
+                    
+        However, :math:`x^{*} = 0` only in the limit of iterations, so in fact this can be infinity.
+        We do not want to have inf values in the convex conjugate, so we have to penalise this value accordingly.
+        The following penalisation is useful in the PDHG algorithm, when we compute primal & dual objectives
+        for convergence purposes.
+        
+        .. math:: F^{*}(x^{*}) = \sum \max\{x^{*}, 0\}
+        
+        """               
+        return x.maximum(0).sum()
+                
+    def proximal(self, x, tau, out=None):
+        
+        """Returns the proximal operator of the constant function, which is the same element, i.e.,
+        
+        .. math:: \mathrm{prox}_{\tau F}(x) = x 
+        
+        """
+        
+        return Identity(x.geometry).direct(x, out=out)
```

Comment:
Isn't this just copying `x` into `out`?

Reply:
yes, but with the line 403, will cover both cases with and without out in one line. Otherwhise, we need to have again if for out = None and out not None

Reply:
yes but doing this will cover in one line both the cases with and without out.